File: LeetcodeProblems/blind-75/matrix/word-search/solution.java
class Solution {
    private boolean[][] visited;
    private int[][] directions = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
    
    public boolean exist(char[][] board, String word) {
        int m = board.length;
        int n = board[0].length;
        visited = new boolean[m][n];
        
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (dfs(board, word, i, j, 0)) {
                    return true;
                }
            }
        }
        return false;
    }
    
    private boolean dfs(char[][] board, String word, int i, int j, int index) {
        if (index == word.length()) {
            return true;
        }
        
        if (i < 0 || i >= board.length || j < 0 || j >= board[0].length || 
            visited[i][j] || board[i][j] != word.charAt(index)) {
            return false;
        }
        
        visited[i][j] = true;
        
        for (int[] dir : directions) {
            if (dfs(board, word, i + dir[0], j + dir[1], index + 1)) {
                return true;
            }
        }
        
        visited[i][j] = false;
        return false;
    }
}

File: LeetcodeProblems/blind-75/matrix/word-search/solution.js
/**
 * @param {character[][]} board
 * @param {string} word
 * @return {boolean}
 */
var exist = function(board, word) {
    const m = board.length;
    const n = board[0].length;
    const visited = Array(m).fill().map(() => Array(n).fill(false));
    const directions = [[-1, 0], [1, 0], [0, -1], [0, 1]];
    
    function dfs(i, j, index) {
        if (index === word.length) {
            return true;
        }
        
        if (i < 0 || i >= m || j < 0 || j >= n || 
            visited[i][j] || board[i][j] !== word[index]) {
            return false;
        }
        
        visited[i][j] = true;
        
        for (const [di, dj] of directions) {
            if (dfs(i + di, j + dj, index + 1)) {
                return true;
            }
        }
        
        visited[i][j] = false;
        return false;
    }
    
    for (let i = 0; i < m; i++) {
        for (let j = 0; j < n; j++) {
            if (dfs(i, j, 0)) {
                return true;
            }
        }
    }
    
    return false;
};
